api/output_policy.go

```diff
@@ -0,0 +1,197 @@
+		return true, nil
+	}
+
+	// Some sudo paths have templated fields in them.
+	// (e.g. sys/revoke-prefix/{prefix})
+	// The keys in the sudoPaths map are actually regular expressions,
+	// so we can check if our path matches against them.
+	for sudoPath := range sudoPaths {
+		r, err := regexp.Compile(fmt.Sprintf("^%s$", sudoPath))
+		if err != nil {
+			continue
+		}
+
+		match := r.Match([]byte(fmt.Sprintf("/%s", path))) // the OpenAPI response has a / in front of each path
+		if match {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func getSudoPaths(client *Client) (map[string]bool, error) {
+	// We don't want to use a wrapping call or any special flags here
+	// so save any custom value and restore after.
+	currentWrappingLookupFunc := client.CurrentWrappingLookupFunc()
+	client.SetWrappingLookupFunc(nil)
+	defer client.SetWrappingLookupFunc(currentWrappingLookupFunc)
+	currentOutputCurlString := client.OutputCurlString()
+	client.SetOutputCurlString(false)
+	defer client.SetOutputCurlString(currentOutputCurlString)
+	currentOutputPolicy := client.OutputPolicy()
+	client.SetOutputPolicy(false)
+	defer client.SetOutputPolicy(currentOutputPolicy)
+
+	r := client.NewRequest("GET", "/v1/sys/internal/specs/openapi")
+	resp, err := client.RawRequest(r)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve sudo endpoints: %v", err)
+	}
+
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body from OpenAPI: %v", err)
+	}
+	if buf.Len() == 0 {
+		return nil, fmt.Errorf("OpenAPI response had no content")
+	}
+
+	oasInfo := make(map[string]interface{})
+	if err := jsonutil.DecodeJSONFromReader(&buf, &oasInfo); err != nil {
+		return nil, fmt.Errorf("unable to decode JSON from OpenAPI response: %v", err)
+	}
+
+	paths, ok := oasInfo["paths"]
+	if !ok {
+		return nil, fmt.Errorf("OpenAPI response did not include paths")
+	}
+
+	pathsMap, ok := paths.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("OpenAPI response did not return valid paths")
+	}
+
+	sudoPaths := make(map[string]bool) // this could be a slice, but we're just making it a map so we can do quick lookups for the paths that don't have any templating
+	for pathName, pathInfo := range pathsMap {
+		pathInfoMap, ok := pathInfo.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		if sudo, ok := pathInfoMap["x-vault-sudo"]; ok {
+			if sudo == true {
+				// Since many paths have templated fields like {name},
+				// our list of sudo paths will actually be a list of
+				// regular expressions that we can match against.
+				pathRegex := buildPathRegexp(pathName)
+				sudoPaths[pathRegex] = true
+			}
+		}
+	}
+
+	return sudoPaths, nil
+}
+
+// Replaces any template fields in a path with the characters ".+" so that
+// we can later allow any characters to match those fields.
+func buildPathRegexp(pathName string) string {
```

Comment:
Ah, I think I understand the code a bit better now, thanks for the detailed explanation!

I think the regex below can still be combined into a single statically-compiled regex:

```go
re := regexp.MustCompile("{path}|{header}|{prefix}|{name}|{type}")
pathWithRegexPatterns = r.ReplaceAllString(pathWithRegexPatterns, ".+")
```

As for constructing a new regex by replacing `{...}` with `.+`, this may run into a few issues. For example, a regex like `a/b/.+` will match not only `a/b/c` but also `a/b/c/d/e/f`, which is probably not what we want here. 

Additionally, the templated name may have characters which would yield the generated regex invalid (e.g. `+`, `*`, '[`, `(`, etc.). All these characters need to be properly escaped if we want to ensure that a valid regex is generated. I would strongly recommend writing a unit test with several test cases to make sure that the escaping logic works.

Perhaps there are other alternatives besides generating regexes (maybe a library that can match the paths as such?)